src/Symfony/Component/Finder/Iterator/VcsIgnoredFilterIterator.php

```diff
@@ -0,0 +1,76 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Finder\Iterator;
+
+use Symfony\Component\Finder\Gitignore;
+
+final class VcsIgnoredFilterIterator extends \FilterIterator
+{
+    /**
+     * @var string
+     */
+    private $baseDir;
+
+    /**
+     * @var array<string, string|null>
+     */
+    private $gitignoreFilesCache = [];
+
+    public function __construct(\Iterator $iterator, string $baseDir)
+    {
+        $this->baseDir = $baseDir;
+
+        parent::__construct($iterator);
+    }
+
+    public function accept(): bool
+    {
+        $file = $this->current();
+
+        $fileRealPath = $file->getRealPath();
+        if ($file->isDir()) {
+            $fileRealPath .= '/';
+        }
+
+        $parentDirectory = $fileRealPath;
+
+        do {
+            $parentDirectory = \dirname($parentDirectory);
+            $relativeFilePath = substr($fileRealPath, \strlen($parentDirectory) + 1);
+
+            $regex = $this->readGitignoreFile("{$parentDirectory}/.gitignore");
+
+            if (null !== $regex && preg_match($regex, $relativeFilePath)) {
+                return false;
+            }
+        } while ($parentDirectory !== $this->baseDir);
```

Comment:
Am I correct that the gitignore files are read over and over again?

If so, please cache the gitignore regex at least one per iterator lifecycle/reset.